<h1>promise.catch</h1> <p>The <strong>catch()</strong> method returns a <code>Promise</code> and deals with rejected cases only. It behaves the same as calling <a href="then" title="The then() method returns a Promise. It takes up to two arguments: callback functions for the success and failure cases of the Promise."><code>Promise.prototype.then(undefined, onRejected)</code></a>.</p> <h2 id="Syntax">Syntax</h2> <pre class="syntaxbox"><var>p.catch(onRejected)</var>;

p.catch(function(reason) {
   // rejection
});
</pre> <h3 id="Parameters">Parameters</h3> <dl> <dt>onRejected</dt> <dd>A <a href="../function" title="The Function constructor creates a new Function object. In JavaScript every function is actually a Function object."><code>Function</code></a> called when the <code>Promise</code> is rejected. This function has one argument: <dl> <dt><code>reason</code></dt> <dd>The rejection reason.</dd> </dl> The Promise returned by <code>catch()</code> is rejected if <code>onRejected</code> throws an error or returns a Promise which is itself rejected; otherwise, it is resolved.</dd> </dl> <h3 id="Return_value">Return value</h3> <p>A <a href="../promise" title="The Promise object is used for asynchronous computations. A Promise represents a value which may be available now, or in the future, or never."><code>Promise</code></a>.</p> <h2 id="Description">Description</h2> <p>The <code>catch</code> method can be useful for error handling in your promise composition.</p> <h2 id="Examples">Examples</h2> <h3 id="Using_and_chaining_the_catch_method">Using and chaining the <code>catch</code> method</h3> <pre data-language="js">var p1 = new Promise(function(resolve, reject) {
  resolve('Success');
});

p1.then(function(value) {
  console.log(value); // "Success!"
  throw 'oh, no!';
}).catch(function(e) {
  console.log(e); // "oh, no!"
}).then(function(){
  console.log('after a catch the chain is restored');
}, function () {
  console.log('Not fired due to the catch');
});

// The following behaves the same as above
p1.then(function(value) {
  console.log(value); // "Success!"
  return Promise.reject('oh, no!');
}).catch(function(e) {
  console.log(e); // "oh, no!"
}).then(function(){
  console.log('after a catch the chain is restored');
}, function () {
  console.log('Not fired due to the catch');
});
</pre> <h3 id="Gotchas_when_throwing_errors">Gotchas when throwing errors</h3> <pre data-language="js">// Throwing an error will call the catch method most of the time
var p1 = new Promise(function(resolve, reject) {
  throw 'Uh-oh!';
});

p1.catch(function(e) {
  console.log(e); // "Uh-oh!"
});

// Errors thrown inside asynchronous functions will act like uncaught errors
var p2 = new Promise(function(resolve, reject) {
  setTimeout(function() {
    throw 'Uncaught Exception!';
  }, 1000);
});

p2.catch(function(e) {
  console.log(e); // This is never called
});

// Errors thrown after resolve is called will be silenced
var p3 = new Promise(function(resolve, reject) {
  resolve();
  throw 'Silenced Exception!';
});

p3.catch(function(e) {
   console.log(e); // This is never called
});</pre> <h2 id="Specifications">Specifications</h2> <table class="standard-table"> <tbody> <tr> <th scope="col">Specification</th> <th scope="col">Status</th> <th scope="col">Comment</th> </tr> <tr> <td><a href="http://www.ecma-international.org/ecma-262/6.0/#sec-promise.prototype.catch" class="external" lang="en" hreflang="en">ECMAScript 2015 (6th Edition, ECMA-262)<br><small lang="en-US">The definition of 'Promise.prototype.catch' in that specification.</small></a></td> <td><span class="spec-Standard">Standard</span></td> <td>Initial definition in an ECMA standard.</td> </tr> <tr> <td><a href="https://tc39.github.io/ecma262/#sec-promise.prototype.catch" class="external" lang="en" hreflang="en">ECMAScript 2017 Draft (ECMA-262)<br><small lang="en-US">The definition of 'Promise.prototype.catch' in that specification.</small></a></td> <td><span class="spec-Draft">Draft</span></td> <td> </td> </tr> </tbody> </table> <h2 id="Browser_compatibility">Browser compatibility</h2>  <div id="compat-desktop"><table class="compat-table"> <thead> <tr> <th>Feature</th> <th>Chrome</th> <th>Edge</th> <th>Firefox</th> <th>Internet Explorer</th> <th>Opera</th> <th>Safari</th> <th>Servo</th> </tr> </thead>
<tbody><tr>
<td>Basic Support</td>
<td>32.0</td>
<td><span style="color: #888;">(Yes)</span></td>
<td>29.0</td>
<td><span style="color: #f00;">No support</span></td>
<td>19</td>
<td>7.1</td>
<td><span style="color: #f00;">No support</span></td>
</tr></tbody>
</table></div><div id="compat-mobile"><table class="compat-table"> <thead> <tr> <th>Feature</th> <th>Android</th> <th>Chrome for Android</th> <th>Edge Mobile</th> <th>Firefox for Android</th> <th>IE Mobile</th> <th>Opera Mobile</th> <th>Safari Mobile</th> </tr> </thead>
<tbody><tr>
<td>Basic Support</td>
<td>4.4.4</td>
<td>32.0</td>
<td><span style="color: #888;">(Yes)</span></td>
<td>29</td>
<td><span style="color: #f00;">No support</span></td>
<td><span style="color: #888;">(Yes)</span></td>
<td>8.0</td>
</tr></tbody>
</table></div> <h2 id="See_also">See also</h2> <ul> <li><a href="../promise" title="The Promise object is used for asynchronous computations. A Promise represents a value which may be available now, or in the future, or never."><code>Promise</code></a></li> <li><a href="then" title="The then() method returns a Promise. It takes up to two arguments: callback functions for the success and failure cases of the Promise."><code>Promise.prototype.then()</code></a></li> </ul><div class="_attribution">
  <p class="_attribution-p">
    <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch$edit" class="_attribution-link">Edit this page on MDN</a>
  </p>
</div>
<div class="_attribution">
  <p class="_attribution-p">
    &copy; 2005&ndash;2017 Mozilla Developer Network and individual contributors.<br>Licensed under the Creative Commons Attribution-ShareAlike License v2.5 or later.<br>
    <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch" class="_attribution-link">https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch</a>
  </p>
</div>
